### PR TITLE
Conformance test redesign

### LINKED ISSUE #416: conformance: make it easier to isolate a single test failure
The conformance tests are not all independent of one another. That is, when there's a failure, isolating that test (for example by changing `Specify` to `FSpecify`) and re-running the test suite will not yield the same test failure, because the tests have interdependencies.

To take one example, I am currently seeing this failure:

```
/home/rogpeppe/other/oci-distribution-spec/conformance/00_conformance_suite_test.go:12
  Push
  /home/rogpeppe/other/oci-distribution-spec/conformance/02_push_test.go:15
    Blob Upload Chunked
    /home/rogpeppe/other/oci-distribution-spec/conformance/02_push_test.go:152
      Get on stale blob upload should return 204 with a range and location [It]
      /home/rogpeppe/other/oci-distribution-spec/conformance/02_push_test.go:215

      Expected
          <int>: 405
      to equal
          <int>: 204

      /home/rogpeppe/other/oci-distribution-spec/conformance/02_push_test.go:220
```

The location of the failure is [here](https://github.com/opencontainers/distribution-spec/blob/7fcdf80f3c8b68e296bad0c5cd19aefe3f7265b1/conformance/02_push_test.go#L220).

If I try to run that test on its own to see exactly what HTTP requests are being sent and received, so to try to understand why this failure is happening and how to fix it, I'd expect to be able to focus on the test by changing line 215 to use `FSpecify`.

However, that doesn't work because on line 217, the test [refers to](https://github.com/opencontainers/distribution-spec/blob/7fcdf80f3c8b68e296bad0c5cd19aefe3f7265b1/conformance/02_push_test.go#L217) `prevResponse`, a variable outside of that test, that has been populated by [this test](https://github.com/opencontainers/distribution-spec/blob/7fcdf80f3c8b68e296bad0c5cd19aefe3f7265b1/conformance/02_push_test.go#L190).

This pattern is common throughout the conformance tests.

In general, it's a good idea for tests to be self-contained. Failing that, it would be nice if tests could at least document their dependencies so running a test on its own doesn't require error-prone tracing of dependencies.

### LINKED ISSUE #501: Additional testing needed to
add another field for contentDiscovery with validation, to test legacy dag mode

see: https://github.com/opencontainers/distribution-spec/pull/491#discussion_r1392914219

### LINKED ISSUE #548: Proposal: refactor conformance tests
The feedback about conformance tests is that they have been useful but UX in general is terrible.

Here are some UX requirements from the last OCI meeting (08/01/2024)

[ ] Root-cause-analysis of test failures should be easier/clearer

[ ] Conformance tests should be configuration (config-file) driven so that
 - the MAY, MUST, SHOULD, etc can be set, and individual tests and suites can be turned on or off
 - the config-file can be shared so that it can be used to verify and derive
 - OCI specs can publish a recommended config-file on GA

[ ] Clients expectations (of registries) from conformance tests

[ ] Registries' expectations (of clients) from conformance tests

[ ] A website revamp for the results/config matrix

### PR DESCRIPTION
**Pull request by sudo-bmitch**
This is a redesign of the conformance tests with the following goals:

- Run as a command instead of a `go test`. This will allow unit testing and avoid some of the Go testing wrappers.
- Make the API testing more granular.
- Design the tests to run against different types of input data, including the various updates to image-spec.
- Minimize dependencies to avoid upstream breaking changes.
- Where possible, note unsupported APIs and fallback to alternatives.

Desired features / tasks include:

- [x] Sparse index tests.
- [x] Edge cases on blob pushes (out of order, chunked push with data in the final put).
- [x] Pushing content with the wrong digest or an invalid digest algorithm.
- [x] Verify all allowed HTTP status codes are appropriately handled.
- [x] Cache/reuse auth tokens between HTTP requests.
- [x] Rename conformance2 to conformance and delete the old conformance tests.
- [ ] ~Update the GitHub actions to test against multiple registries, and consider if/when failing 3rd party registries should block a merge to the distribution-spec.~ (Infeasible since GHA doesn't have a warning status for the results.)
- [x] Generate an image in GitHub actions to be used by a conformance action.
- [x] Update the conformance action.
- [x] Update the oci-conformance repo to handle the new input (results will likely be summarized with two columns, API and Data, with green or gray status to show passing tests or unsupported). (Implemented in https://github.com/opencontainers/oci-conformance/pull/130)

Other wish list items:

- [ ] Add some unit tests.
- [x] Add ability to run a single test.
- [ ] Add separate tests for foreign layers (URL defined) from non-distributable layers (specific media type).

Merging this will unblock #543. Closes #548. Closes #501. Closes #416.

### CODE REVIEW

File: conformance2/run.go [hunk 1]
```diff
@@ -0,0 +1,1684 @@
...
+		dataTests = append(dataTests, tdName)
+		_, err := r.State.Data[tdName].genManifestFull(
+			genWithTag("data-field"),
+			genWithDescriptorData(),
+		)
+		if err != nil {
+			return fmt.Errorf("failed to generate test data: %w", err)
+		}
+	}
+	// image with non-distributable layers
```
Comment (tianon) [inline conformance2/run.go]: These tests appear to be conflating two concepts :innocent:

For lack of an officially spec-defined term, blobs whose descriptors include the `urls` key are colloquially known as "foreign layers", and technically any blob _could_ be foreign (with zero change to the `mediaType`), although most registries probably validate `config` differently from `layers` in that respect (I know Docker Hub certainly does :innocent:).

Non-distributable layers (those with `mediaType` of `application/vnd.oci.image.layer.nondistributable.v1.tar*`), on the other hand, may or may not be explicitly "foreign" (ie, they may or may not have `urls` set).

- https://github.com/opencontainers/image-spec/blob/v1.1.1/descriptor.md#:~:text=This%20OPTIONAL%20property%20specifies%20a%20list%20of%20URIs%20from%20which%20this%20object%20MAY%20be%20downloaded.

  > This OPTIONAL property specifies a list of URIs from which this object MAY be downloaded.

- https://github.com/opencontainers/image-spec/blob/v1.1.1/media-types.md#:~:text=The%20following%20media%20types%20identify%20a%20%22Layer%22%20with%20distribution%20restrictions

- https://github.com/opencontainers/image-spec/blob/v1.1.1/layer.md#non-distributable-layers

  > Descriptors referencing non-distributable layers MAY include `urls` for downloading these layers directly; however, the presence of the `urls` field SHOULD NOT be used to determine whether or not a layer is non-distributable.

(For historical curiosity, this evolved/matured/split in two parts from the `application/vnd.docker.image.rootfs.foreign.diff.tar.gzip` media type in the previous "Docker Image Manifest Version 2, Schema 2" spec: https://github.com/distribution/distribution/blob/10e4312e62ab3b0c53bc0e31ade82fbe38af1b43/docs/content/spec/manifest-v2-2.md#:~:text=application/vnd.docker.image.rootfs.foreign.diff.tar.gzip%3A%20%22Layer%22%2C%20as%20a%20gzipped%20tar%20that%20should%20never%20be%20pushed -- this is where I've cribbed the term "foreign layer" from, because "blob/descriptor with `urls` set" is a mouthful.)

Comment (tianon) [reply conformance2/run.go]: Yeah, as I clarified on the call yesterday (and writing here for posterity), I think I'm suggesting we should probably test all the permutations (urls and not urls, pushed blob and not pushed blob, and regular and nondistributable media types for both variances).

**Comment by sudo-bmitch** [comment]
@opencontainers/distribution-spec-maintainers while looking at the options for reusing auth tokens between requests, I'm seeing the following options:

1. Don't reuse tokens. This is the current design, matches the former conformance tests, is the most likely to work, but can be very slow and does not match how clients access registries today.
2. Generate the scope on the client side and reuse the token when the scope is unchanged. I think this matches some of the original docker implementations, but I'm not a fan of this client side logic since that goes against the upstream RFCs.
3. Cache the token response and always try to reuse it, but discard it and regenerate on any auth failure. I worry this could lead to bouncing tokens when things go from get to put to delete requests. And if a registry returns a 403 instead of a 401 when the old token is reused, we either fail or need to run 3 requests, one with a failing token to get a 403, one with no token to get the 401 + auth header, and a third with the valid token.
4. Cache the token for a "transaction" where the transaction may be multiple get requests, or a full push workflow for the layers and manifest. I think this most closely aligns with clients. However with a head followed by a push it can result in scope changes between the requests, and it assumes all clients have the same "transaction" concept.
5. Cache the token per repository + http method. This would allow the most reuse of tokens between tests, but I worry it doesn't match existing clients, and gets back into client side logic that doesn't match the upstream RFCs.

Is there a preference for which option we should use for the conformance test (at least pending the outcome of wg-auth)?

**Comment by jcarter3** [comment]
I _like_ **Option 5** the most, but I'm not sure if it's completely safe.

Given that these tests are for the dist spec (and there is no auth spec), I'm not terribly concerned about trying to match client behavior around auth, and so if **Option 5** let's us reuse tokens the most and do these test faster, I'm all for it.

That said, since there is no auth spec... I know this approach works for Docker Hub and several other registries, but it's not guaranteed that every registry generates tokens the same way off of repository and http method, so it _could_ introduce unrelated errors. Because of that, **Option 4** is probably the safest from a compatibility standpoint, and should at least provide marginal re-use.

**Comment by sudo-bmitch** [comment]
I've added option 5 for caching the auth, with an option to disable it and revert to option 1. If that causes too many issues, we can try implementing option 4.

File: conformance/.gitignore [hunk 2]
```diff
@@ -1,6 +1,7 @@
-vendor/
-junit.xml
-report.html
+!go.mod
+!go.sum
```
Comment (sudo-bmitch) [inline conformance/.gitignore]: If we can get a merge on #601, I'd like to remove these.

File: conformance/go.mod [hunk 3]
```diff
@@ -1,26 +1,10 @@
-module github.com/opencontainers/distribution-spec/conformance
+module github.com/opencontainers/distribution-spec/conformance2
```
Comment (sudo-bmitch) [inline conformance/go.mod]: fix this path

File: conformance/run.go [hunk 4]
```diff
@@ -0,0 +1,2433 @@
...
+			opts = append(opts, apiWithFlag("RequireDigestHeader"))
+		}
+		if err := r.API.ManifestPut(r.Config.schemeReg, repo, dig.String(), dig, td, r.Config.APIs.Referrer, putOpts, opts...); err != nil {
+			r.TestFail(res, err, tdName, apis...)
+			return fmt.Errorf("%.0w%w", errAPITestFail, err)
+		}
+		if len(expectTags) > 0 {
+			errs := []error{}
+			for _, tag := range expectTags {
+				if slices.Contains(resp.Header.Values("OCI-Tag"), tag) {
```
Comment (tianon) [inline conformance/run.go]: There's a tiny bug here - https://github.com/opencontainers/distribution-spec/pull/600 specifies that `OCI-Tag` can be comma-separated, but this code as-written will only accept multiple headers. :smile:

See https://github.com/distribution/distribution/pull/1782 for where I fixed a similar bug for `Accept` headers in dist/dist eons ago :joy:

Comment (tianon) [reply conformance/run.go]: There's probably a cleaner way to write it, but something like this:

```suggestion
				if slices.ContainsFunc(resp.Header.Values("OCI-Tag"), func(v string) bool {
					return slices.ContainsFunc(strings.Split(v, ","), func(s string) bool {
						return strings.TrimSpace(s) == tag
					})
				}) {
```

Comment (sudo-bmitch) [reply conformance/run.go]: Thanks for catching that. I've pushed your fix.

**Review by dmcgowan — APPROVED** [review]

**Review by tianon — APPROVED** [review]

**Review by mikebrow — APPROVED** [review]
LGTM
